ext/dcos-installer/dcos_installer/action_lib.py

```diff
@@ -43,27 +43,73 @@ class ExecuteException(Exception):
     """Raised when execution fails"""
 
 
-def nodes_count_by_type(config):
-    total_agents_count = len(config.get('agent_list', [])) + len(config.get('public_agent_list', []))
+def get_successful_nodes_from(step: str, state_json_dir: str) -> list:
+    assert step in ['preflight', 'deploy', 'postflight']
+    step += '.json'
+
+    step_path = os.path.join(state_json_dir, step)
+    step_json = _read_state_file(step_path)
+    if not step_json:
+        # TODO: Fix exception here
+        raise Exception("Could not get hosts from {}".format(step_path))
+
+    assert 'hosts' in step_json, "Field `hosts` not found, location {}".format(step_path)
+
+    nodes = []
+    for node_ip, node_params in step_json['hosts'].items():
+        assert 'host_status' in node_params, "Field `host_status` not found for node {}".format(node_ip)
+        assert 'tags' in node_params, 'Field `tags` not found for node {}'.format(node_ip)
+        if node_params['host_status'] == 'success':
+            nodes.append(Node(node_ip, tags=node_params['tags']))
+        else:
+            assert 'role' in node_params['tags'], 'tags must contain parameter role. Got {}'.format(node_params['tags'])
+            if node_params['tags']['role'] == 'master':
+                raise Exception("Master node {} has failed. Uable to continue".format(node_ip))
```

Comment:
Pretty sure the Web installer code won't be happy with this currently unfortunately.